File: frontend/src/compiler/keygen.kat.test.ts
// X25519 known-answer test (KAT) — THE crypto-correctness gate for the TS keygen seam. A wrong-but-
// plausible public key would pass every shape check yet fail a real WireGuard handshake on hardware, so
// this pins @noble/curves x25519 derivation byte-for-byte against the Go oracle's vectors.
//
// Source of truth: internal/conformance/testdata/keygen_kat.json (Go-authored, RFC 7748 section 6.1 +
// clamp-boundary vectors). public_b64 is the base64.StdEncoding X25519 public key, proven byte-equal to
// the default wgtypesKeygen over 10k inputs by keygen_equivalence_test.go on the Go side. Asserting
// @noble against THIS file makes plans 3/4/5 all pin one X25519 derivation definition.
//
// The clamp_equivalence vectors prove @noble's INTERNAL clamp matches Go: a raw private key and its
// X25519-clamped form (key[0] &= 248; key[31] &= 127; key[31] |= 64) MUST derive the identical public
// key. This is the property the seam relies on by calling getPublicKey (which clamps) WITHOUT a hand
// pre-clamp — a double-clamp or a skipped-clamp port would fail here.

import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, it } from 'vitest';

import { derivePublic, parseAndNormalize } from './keygen';

// thisDir = frontend/src/compiler; three '..' hops reach the repo root (compiler → src → frontend →
// root), mirroring the heal.conformance.test.ts loader idiom so both read the shared Go corpus the same
// way.
const thisDir = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(thisDir, '..', '..', '..');
const katPath = join(repoRoot, 'internal/conformance/testdata/keygen_kat.json');

interface KatVector {
  name: string;
  doc: string;
  private_b64: string;
  public_b64: string;
}

interface ClampEquivVector {
  name: string;
  doc: string;
  raw_b64: string;
  clamped_b64: string;
  public_b64: string;
}

interface KatFile {
  note: string;
  vectors: KatVector[];
  clamp_equivalence: ClampEquivVector[];
}

const kat = JSON.parse(readFileSync(katPath, 'utf8')) as KatFile;

describe('keygen X25519 KAT (Go-authored RFC 7748 + clamp-boundary vectors)', () => {
  // Sanity: the corpus loaded and is non-empty, so a silently-empty fixture can't make the suite a
  // vacuous pass.
  it('loaded the Go KAT corpus with vectors', () => {
    expect(kat.vectors.length).toBeGreaterThan(0);
    expect(kat.clamp_equivalence.length).toBeGreaterThan(0);
  });

  // Every vectors[] entry: derivePublic(private_b64) === public_b64. This is the core assertion that
  // @noble's getPublicKey reproduces the Go/crypto-ecdh public key byte-for-byte.
  for (const v of kat.vectors) {
    it(`derivePublic matches Go for ${v.name}`, () => {
      expect(derivePublic(v.private_b64)).toBe(v.public_b64);
    });
  }

  // Every clamp_equivalence[] entry: BOTH the raw and the pre-clamped private key derive the SAME
  // expected public key, proving @noble clamps internally exactly as Go does (so the seam correctly does
  // not pre-clamp).
  for (const v of kat.clamp_equivalence) {
    it(`raw and clamped derive identical public key for ${v.name}`, () => {
      expect(derivePublic(v.raw_b64)).toBe(v.public_b64);
      expect(derivePublic(v.clamped_b64)).toBe(v.public_b64);
    });
  }
});

describe('keygen parseAndNormalize round-trip', () => {
  // FAITHFULNESS to Go's StdEncoding: an UNPADDED 32-byte key is NOT valid base64.StdEncoding — Go's
  // base64.StdEncoding.DecodeString rejects it ("illegal base64 data") because the encoded length is not
  // a multiple of 4. A lenient atob would silently decode it, so parseAndNormalize MUST throw to stay
  // byte-exact to wgtypes.ParseKey (no TS-accepts-where-Go-rejects). This is the flipped assertion: the
  // unpadded variant is a REJECTION, not a canonicalizable input.
  it('throws on an unpadded 32-byte key (Go StdEncoding rejects it)', () => {
    const canonical = kat.vectors[0].private_b64; // 44-char padded base64-std of 32 bytes
    const unpadded = canonical.replace(/=+$/, ''); // strip the trailing '=' pad: NOT valid StdEncoding
    expect(unpadded).not.toBe(canonical);
    expect(() => parseAndNormalize(unpadded)).toThrow();
  });

  // The ONE non-canonical encoding Go's StdEncoding DOES accept is one carrying embedded '\n' / '\r':
  // the decoder skips those bytes. parseAndNormalize must decode such a variant and re-emit the
  // newline-free canonical form, identical to wgtypes.ParseKey(priv).String() — proving the seam matches
  // Go's newline tolerance (no TS-rejects-where-Go-accepts).
  it('re-encodes a newline-laden key to the canonical newline-free base64-std form', () => {
    const canonical = kat.vectors[0].private_b64;
    // Splice a '\n' and a '\r' into the middle — Go ignores both; the canonical output carries neither.
    const withNewlines =
      canonical.slice(0, 16) + '\n' + canonical.slice(16, 32) + '\r' + canonical.slice(32);
    expect(withNewlines).not.toBe(canonical);
    expect(parseAndNormalize(withNewlines)).toBe(canonical);
  });

  // A space-laden key is REJECTED: Go's StdEncoding skips ONLY '\n'/'\r', never a space (or tab). This
  // pins the asymmetry that the BLOCKER fix introduced — space is not whitespace the decoder tolerates.
  it('throws on a space-laden key (Go StdEncoding rejects space)', () => {
    const canonical = kat.vectors[0].private_b64;
    const withSpace = canonical.slice(0, 16) + ' ' + canonical.slice(16);
    expect(() => parseAndNormalize(withSpace)).toThrow();
  });

  // An already-canonical key round-trips to itself (idempotence), and the normalized output still derives
  // the same public key as the original — the air-gap case-a invariant that normalize-then-derive is
  // stable.
  it('is idempotent on an already-canonical key and preserves the derived public key', () => {
    const canonical = kat.vectors[0].private_b64;
    expect(parseAndNormalize(canonical)).toBe(canonical);
    expect(derivePublic(parseAndNormalize(canonical))).toBe(kat.vectors[0].public_b64);
  });
});

File: internal/conformance/keygen_kat_test.go
package conformance

import (
	"crypto/ecdh"
	"encoding/base64"
	"encoding/json"
	"os"
	"testing"
)

// keygen_kat_test.go — X25519 known-answer test (step 4 of plan-5 / 1.5).
//
// It pins the WireGuard key-derivation definition the whole conformance program rests on:
// the public key is X25519(private, basepoint), base64.StdEncoding-encoded. The test derives
// public keys here by calling crypto/ecdh X25519 DIRECTLY — the SAME stdlib derivation
// internal/localcompile.ecdhKeygen.DerivePublic wraps. ecdhKeygen is unexported (a deliberate
// reference/test implementation not wired into production), so the KAT cannot call it across the
// package boundary; calling ecdh.X25519().NewPrivateKey(raw).PublicKey() reproduces its derivation
// byte-for-byte (ecdhKeygen IS that one line plus base64). plan-3 proved ecdhKeygen byte-equal to
// the default wgtypesKeygen (the production path) over 10k inputs, so this single definition is the
// anchor plans 3, 4, and 5 all pin: the harness here, plan-4's @noble/curves port (which asserts
// against testdata/keygen_kat.json), and the live pipeline.
//
// crypto/ecdh's X25519 clamps the scalar internally during the multiplication (exactly as
// curve25519.ScalarBaseMult does inside wgtypes' Key.PublicKey), so an UNCLAMPED private key and
// its clamped form derive the identical public key — the clamp_equivalence group asserts that
// boundary property a TS port must reproduce.

// keygenKAT is the on-disk testdata/keygen_kat.json shape.
type keygenKAT struct {
	Note    string `json:"note"`
	Vectors []struct {
		Name       string `json:"name"`
		Doc        string `json:"doc"`
		PrivateB64 string `json:"private_b64"`
		PublicB64  string `json:"public_b64"`
	} `json:"vectors"`
	ClampEquivalence []struct {
		Name       string `json:"name"`
		Doc        string `json:"doc"`
		RawB64     string `json:"raw_b64"`
		ClampedB64 string `json:"clamped_b64"`
		PublicB64  string `json:"public_b64"`
	} `json:"clamp_equivalence"`
}

// derivePublicB64 reproduces internal/localcompile.ecdhKeygen.DerivePublic: decode the base64
// X25519 private key, derive the public half via crypto/ecdh (which clamps internally), re-encode
// base64.StdEncoding. Keeping this one-liner local (rather than importing the unexported ecdhKeygen)
// is what makes the KAT a SECOND independent witness of the derivation — if ecdhKeygen ever drifts
// from the X25519 definition, the equivalence test in localcompile catches it there and this KAT
// catches it here.
func derivePublicB64(t *testing.T, privB64 string) string {
	t.Helper()
	raw, err := base64.StdEncoding.DecodeString(privB64)
	if err != nil {
		t.Fatalf("decode private key %q: %v", privB64, err)
	}
	priv, err := ecdh.X25519().NewPrivateKey(raw)
	if err != nil {
		t.Fatalf("build X25519 private key from %q: %v", privB64, err)
	}
	return base64.StdEncoding.EncodeToString(priv.PublicKey().Bytes())
}

func loadKAT(t *testing.T) keygenKAT {
	t.Helper()
	raw, err := os.ReadFile("testdata/keygen_kat.json")
	if err != nil {
		t.Fatalf("read keygen_kat.json: %v", err)
	}
	var kat keygenKAT
	if err := json.Unmarshal(raw, &kat); err != nil {
		t.Fatalf("parse keygen_kat.json: %v", err)
	}
	if len(kat.Vectors) == 0 {
		t.Fatal("keygen_kat.json has no vectors")
	}
	if len(kat.ClampEquivalence) == 0 {
		t.Fatal("keygen_kat.json has no clamp_equivalence vectors")
	}
	return kat
}

// TestKeygenKAT is the known-answer test: every vector's private key must derive its frozen public
// key, and every clamp-equivalence pair (a raw private key and its X25519-clamped form) must derive
// the SAME public key. RFC 7748 section 6.1's Alice/Bob vectors anchor the derivation against the
// authoritative IETF answer; the clamp-boundary vectors exercise key[0] &= 248, key[31] &= 127, and
// key[31] |= 64.
func TestKeygenKAT(t *testing.T) {
	kat := loadKAT(t)

	t.Run("vectors", func(t *testing.T) {
		for _, v := range kat.Vectors {
			v := v
			t.Run(v.Name, func(t *testing.T) {
				got := derivePublicB64(t, v.PrivateB64)
				if got != v.PublicB64 {
					t.Errorf("X25519 public-key mismatch for %s:\n  priv: %s\n  want: %s\n  got:  %s",
						v.Name, v.PrivateB64, v.PublicB64, got)
				}
			})
		}
	})

	t.Run("clamp_equivalence", func(t *testing.T) {
		for _, c := range kat.ClampEquivalence {
			c := c
			t.Run(c.Name, func(t *testing.T) {
				rawPub := derivePublicB64(t, c.RawB64)
				clampedPub := derivePublicB64(t, c.ClampedB64)
				if rawPub != clampedPub {
					t.Errorf("clamp equivalence broken for %s: raw and clamped private keys derived DIFFERENT public keys (X25519 must clamp internally)\n  raw_pub:     %s\n  clamped_pub: %s",
						c.Name, rawPub, clampedPub)
				}
				if rawPub != c.PublicB64 {
					t.Errorf("clamp equivalence %s: derived public %s does not match the frozen public %s",
						c.Name, rawPub, c.PublicB64)
				}
			})
		}
	})
}
